src/lazy/expanded/macro_evaluator.rs

```diff
@@ -642,31 +750,226 @@ mod tests {
     ///
     /// This test exists to demonstrate that macro evaluation within the TDL context works the
     /// same as evaluation in the data stream.
-    fn eval_tdl_template_invocation(invocation: &str, expected: &str) -> IonResult<()> {
-        let macro_table = MacroTable::new();
-        let symbol_table = SymbolTable::new();
-        let allocator = BumpAllocator::new();
-        let context = EncodingContext::new(&macro_table, &symbol_table, &allocator);
-        let mut evaluator = TdlMacroEvaluator::<TextEncoding_1_1>::new();
-        let invocation = Element::read_one(invocation)?;
-        let actuals = evaluator.evaluate(context, invocation.expect_sexp()?)?;
+    fn eval_template_invocation(
+        template_definition: &str,
+        invocation: &str,
+        expected: &str,
+    ) -> IonResult<()> {
+        let mut reader = LazyTextReader_1_1::new(invocation.as_bytes())?;
+        let _macro_address = reader.register_template(template_definition)?;
+        let actuals = reader.read_all_elements()?;
         let mut expected_reader = LazyTextReader_1_1::new(expected.as_bytes())?;
-        for actual_result in actuals {
+        for actual in actuals {
             // Read the next expected value as a raw value, then wrap it in an `ExpandedRawValueRef`
             // so it can be directly compared to the actual.
             let expected: Element = expected_reader.next()?.unwrap().read()?.try_into()?;
-            let actual: Element = LazyValue::from(actual_result?).try_into()?;
             assert_eq!(actual, expected);
         }
         assert!(matches!(expected_reader.next(), Ok(None)));
 
         Ok(())
     }
 
+    #[test]
+    fn multiple_top_level_values() -> IonResult<()> {
+        eval_template_invocation(
+            "(macro foo () (values 1 2 3 4 5))",
+            r#"
+                (:foo)
+            "#,
+            r#"
+                1 2 3 4 5
+            "#,
+        )
+    }
+
+    #[test]
+    fn emit_symbol_table() -> IonResult<()> {
+        eval_template_invocation(
+            r#"
+                (macro lst (symbols) 
+                    $ion_symbol_table::{
+                        symbols: symbols
+                    }
+                )
+            "#,
+            r#"
+                (:lst ["foo", "bar", "baz"]) $10 $11 $12
+            "#,
+            r#"
+                foo bar baz
+            "#,
+        )
+    }
+
+    #[test]
+    fn context_changes_happen_between_top_level_expressions() -> IonResult<()> {
```

Comment:
🗺️ The spec requires that a single expression use the same encoding for its entirety. If an expression produces multiple values, any system values will not be applied until the expression has been evaluated in full.